### DISCUSSION TITLE
Running on Macbook M1

### DISCUSSION THREAD

**Discussion by jefharris** [lllyasviel/stable-diffusion-webui-forge#270] [Q&A]
Anyone successfully installed and run it on a Macbook M1?

**Comment by barepixels** [comment]
I have a Macbook M1 but never do Stable Diffusion on it.  Am curious how long in seconds to render a 1024x1024 on it

**Comment by humanplugged** [comment]
I did try to install it on a M1, but it's a bit confusing, at least for me...

A quick tutorial would be very grateful! ♥️

(I do have Automatic1111 working 100% locally, also Pinokio app using a few other AI tools, even tough it woks well, it's quite heavy on my M1 Max)

**Reply by barepixels** [reply]
how long in seconds to render a 1024x1024 on M1 Max

**Reply by humanplugged** [reply]
Hey barepixels.

with these arguments: --skip-torch-cuda-test --skip-install --no-half-vae --upcast-sampling --precision full

with some SDXL turbo models, Euler a , steps 6, cfg 2, 1024x1024,   rendering is about 18.1 sec 

with LCM's , lcm, steps 4, cfg 1,5, 1024x1024, rendering is about 14.2 sec.

it can get a bit faster using SDXL turbo models by lowering the steps. EG: step of 4 it's about 11.2 sec.

I'm quite curious to see what difference it could get from Forge.

**Comment by MonetEinsley** [comment]
I got it working on my MacBook Air M1 (8GB)
If you’re talking about how to install it, I simply entered “git clone https://github.com/lllyasviel/stable-diffusion-webui-forge.git“ in the Terminal window. (Of course, I believe you have to use homebrew or something to get git installed, there should be some videos about that on YouTube https://www.youtube.com/watch?v=Jh-clc4jEvk)

After that, and since I also have an Automatic 1111 installation, I edited the webui-user.sh file in the Forge folder at line 13:
A1111_Home=/Users/moneteinsley/stable-diffusion-webui
export COMMANDLINE_ARGS=" --skip-torch-cuda-test
--ckpt-dir $A1111_Home/models/Stable-diffusion
--hypernetwork-dir $A1111_Home/models/hypernetworks
--embeddings-dir $A1111_Home/embeddings
--lora-dir $A1111_Home/models/Lora
--vae-dir $A1111_Home/models/VAE"

Obviously, make sure you’re using your username and not mine XD
These directory pointers allow me to simply use the models/checkpoints/etc from my Automatic 1111 folder without having to copy them.

Then, whenever I want to run forge, I open up the Teriminal window, enter “cd stable-diffusion-webui-forge”, “git pull” to update it, and “./webui.sh” to run it.

That being said, I haven’t seen any significant difference in terms of performance using SD1.5 between A1111 and Forge. It takes about 10-16 minutes for me to generate 1 image using an LCM checkpoint. (8-16 steps for the first pass at 512x512, 8 steps for hires fix to 1024x1024, and 8 steps for Adetailer face fixes at 512x512). But on the bright side, I’m finally able to actually use hires fix like how I do it on my PC by using Forge.

The problem for me though, is that I’m currently unable to use SDXL checkpoints on Forge. It keeps giving me a TypeError: ‘NoneType’ object is not iterable
No idea how to go about solving that at the moment. Oddly enough, SDXL does kinda work for me on A1111, but only on low resolutions like 512x512.

I’m sure there might be a better way to optimize it for MacOS, but I simply don’t know at this point in time.

Edit: Thanks to dermesut for pointing this out, but you can further simplify the webui-user.sh file to something like this:

A1111_Home=/Users/moneteinsley/stable-diffusion-webui
export COMMANDLINE_ARGS=" --skip-torch-cuda-test
--forge-ref-a1111-home $A1111_Home
--gfpgan-model $A1111_Home/models/GFPGAN
"
And that will basically allow you to use your models/controlnets, etc from your A1111 folder for use in Forge

**Reply by JesusVScode** [reply]
I have an M2 pro and recently installed Forge. My models keep unloading and loading after every action and because of this, it is taking significantly longer to generate images. In my terminal it said "Memory set: Shared" does this have something to do with it or there's something else I should change?

**Reply by MonetEinsley** [reply]
Yeah, I’ve noticed that too, the loading and unloading stuff that is. I honestly don’t quite know what’s going on with that, but I believe that’s supposed to happen, and probably happens all the time in A1111. In other words, I have a suspicion that Forge simply reports every time it loads/unloads the models, whereas A1111 simply doesn’t tell you. 

I believe this could be the case, because I’m sure on ComfyUI in order to upscale an image to do the equivalent of hires fix, it has to encode the image into a latent, upscale the latent, then decode the latent into an image for the model to process. It would make sense then for the models to be unloaded while that occurs, to save VRAM. I could be very wrong on that, as it’s just my guess based on what I’ve seen on ComfyUI, so don’t take my word for it.

I don’t know about “Memory set: Shared”, but my console after it looks at my total VRAM/RAM (8192 MB), it says “Set vram state to: SHARED”. Again, I couldn’t tell you what that’s about—and it’s probably working as intended in that regard (the vram state thing, that is), but if this Memory set thing is basically the same thing, then I doubt it has much to do with the directory pointers, if that was what you’re asking. 

Basically, what I’m saying here is I believe you’ll see the same thing if you migrated your models from your A1111 folders to the Forge ones (and removing the directory pointer arguments from the webui-user.sh file.) I’ve glanced through the settings, but haven’t seen any obvious things that would increase the performance in terms of generation speed. Again, don’t take my word for it, as I’m quite new to Forge myself.

**Reply by AlanDrang** [reply]
I'm using an M1 Max computer, and I modified these lines of code in the webui-user.sh file, but when I open it, it shows an error. Could you help me with that?
![Snipaste_2024-05-07_17-23-15](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/169144197/47e544e0-3da1-4e27-a25b-6258a6fba270)
![Snipaste_2024-05-07_17-26-21](https://github.com/lllyasviel/stable-diffusion-webui-forge/assets/169144197/aa585480-2823-4f63-9f42-d93448410a84)

A1111_Home=/Users/Alan/stable-diffusion-webui
export COMMANDLINE_ARGS=" --skip-torch-cuda-test
--forge-ref-a1111-home $A1111_Home
--gfpgan-model $A1111_Home/models/GFPGAN
"
